src/plugins/saved_objects_management/common/index.ts

```diff
@@ -14,4 +14,19 @@ export type {
   SavedObjectInvalidRelation,
   SavedObjectGetRelationshipsResponse,
   SavedObjectManagementTypeInfo,
+  BulkGetBodyHTTPV1,
+  BulkGetResponseHTTPV1,
+  BulkDeleteBodyHTTPV1,
+  BulkDeleteResponseHTTPV1,
+  FindQueryHTTPV1,
+  FindResponseHTTPV1,
+  FindSearchOperatorHTTPV1,
+  FindSortOrderHTTPV1,
+  GetAllowedTypesResponseHTTPV1,
+  ReferenceHTTPV1,
+  RelationshipsParamsHTTPV1,
+  RelationshipsQueryHTTPV1,
+  RelationshipsResponseHTTPV1,
+  ScrollCountBodyHTTPV1,
+  ScrollCountResponseHTTPV1,
```

Comment:
So,

The PR looks fine to me.

My major question is regarding how we want to manage versioned types. We're more or less introducing a precedent on what best practices should be here, given the SOM preparation is supposed to be later eventually used as an example for other teams.

Using a suffix to specify the version, as done in the PR (and already in some of Kibana's codebase) would for sure be an option, but I wonder if that would be the best one.

Just speaking out loud here, but I wonder if mimicking the way go imports works could eventually be more explicit, while letting us doing less work and/or refactoring.

What I have in mind would be something like that

```ts
import * as apiV1 from `somewhere/v1`;
import * as apiV2 from `somewhere/v2`;

// later

const someResponse: apiV1.RelationshipsResponse = ...;
```

Or, using named imports:

```ts
import { RelationshipsResponse as RelationshipsResponseV1 } from `somewhere/v1`;
import { RelationshipsResponse as RelationshipsResponseV2 } from `somewhere/v2`;
```

Do you think that could somehow make sense?

Reply:
> My major question is regarding how we want to manage versioned types.

Yeah 👍🏻 definitely the thing we should try to figure out in this PR.

> import * as apiV2 from `somewhere/v2`;

Personally, I really like ☝🏻 pattern of namespacing imports (so this would be my first choice). But I don't see much precedent of using `* as myNamespace` imports so I chose to avoid it. Happy to use `* as` if others also like this pattern too.

> import { RelationshipsResponse as RelationshipsResponseV1 } from `somewhere/v1`;

For me, this achieves the same thing, but a bit more verbose. Which is why my preference goes to `* as ...`.

> while letting us doing less work and/or refactoring

I was thinking about this in broadly two use cases:

1. Code that does not need to know about different versions (perhaps most app code?)
2. Code that does need to know about different versions (route handlers, transformations)

For (1), IMO, the best way to reduce refactoring would be to use a mechanism like:

```ts
// v1.ts
interface Foo {}

// shared_types.ts
import * as v1 from './v1.ts'
export type Foo = v1.Foo;
```

Code uses `Foo` type and when `FooV2` comes around you can just point the alias at `FooV2` and the rest of your refactor efforts just need to make sure your code uses the new `Foo` correctly. Additionally, with this pattern much of your app code would not need to use namespacing techniques except where absolutely necessary.

For (2) code needs a pattern for working with 1 or more versioned types. I think route handlers will definitely need this because they will themselves be versioned. Hopefully this kind of code can be "pushed to edges" or otherwise contained so that Kibana code does become littered with `if (isV1(myObject)) {} else if (isV2(myObject)) {}`. That being said, something like:

```ts
import * as v1 from 'types/v1';
...
  const body: v1.GetFooHTTP =  { ... };
  return res.ok({ body });
...
```

---

One other thought I had was that it would be nice if our code can "guide" us to correctly evolve a schema and all other schemas that reference it. Using the `* as ...` pattern this might look something like:

```
// v1.ts
interface Foo {}

interface GetFooResponseHTTP { foo: Foo }

// introduce new field in Foo => v2.ts
import * as v1 from './v1'

interface Foo extends v1.Foo {newField: string}

interface GetFooResponseHTTP { foo: Foo }
```

Reply:
> just speaking out loud here, but I wonder if mimicking the way go imports works could eventually be more explicit, while letting us doing less work and/or refactoring.

++ having a generic `somewhere` to import types from opens up possibilities of having these types live in type-specific packages (although, we could do that right now anyway). We could take that even further and eventually implement something similar to how ecs types are now [automatically updated](https://github.com/elastic/kibana/pull/148804) on each release.
I digress though.

> Personally, I really like ☝🏻 pattern of namespacing imports (so this would be my first choice). But I don't see much precedent of using * as myNamespace imports so I chose to avoid it. Happy to use * as if others also like this pattern.

FWIW, my preference goes towards being verbose and _explicitly_ importing only what one needs. It takes the "guess" work out of what we're actually importing.

Reply:
> Personally, I really like ☝🏻 pattern of namespacing imports 

I really like it too tbh. Having used the pattern for k8s development and versioned APIs, it's efficient and KISS at the same time ihmo.

However I agree that it's not commonly used in our codebase (`*` imports were/are even discouraged because of usages with concrete imports), so I won't really object if we think it's not suitable. 

I still makes so much sense for type-only imports though... 

> For me, this achieves the same thing, but a bit more verbose

Definitely not as good. I was mostly highlighting the fact that it would still be possible to do unitary imports if necessary (no `import *`)

> One other thought I had was that it would be nice if our code can "guide" us to correctly evolve a schema and all other schemas that reference it.

I get the idea, and see the upsides in maintainability and code re-use. We still have to keep in mind that there are a few major downsides doing this kind of `v2 extends v1` patterns:

1. Worse UX: forces to dig into nested types to see your structure
2. Related to the previous point - it also makes generated documentation worse
3. It forces to move all the things around to remove a version (e.g when you want to delete version 1, you're forced to either keep the associated types as they're referenced by newer version, or to refactor your types)

So overall I'm not sure yet the upsides outweights the downsides?

> For (1), IMO, the best way to reduce refactoring would be to use a mechanism like:

Why not just 

```
// latest.ts

export * from '{current}.ts'
```

so that consumers could just use

```ts
import  { MyType } from 'somewhere/api/latest'
```

Reply:
I'd be +1 on namespace imports, we also use it for `fp-ts` imports and I like that it makes the version explicit in e.g. the route handler.

> * imports were/are even discouraged because of usages with concrete imports

@pgayvallet I'm not sure what you mean here?

I also like the `latest` "symlink" :+1:
I'm aware we had problems with `export * from './...'` because of api-extractor. Some of the api-extractor problems might also affect the typescript tooling we have to find usages, deprecated APIs or use of `any` but we'd have to dig to try to find out. But this pattern can work even with explicitly mentioning all the types, it's just mork work to keep in sync.

Reply:
> Why not just // latest.ts

I assume that interfaces won't necessarily all have new versions at the same rate. So you might have something like:

```ts
// v1.ts

export interface Foo {}

export interface GetFooHTTP { foo: Foo }

export interface Bar {}

// v2.ts

export interface Foo { newField: string }

export interface GetFooHTTP { foo: Foo }

// latest.ts

// Now "latest" would be missing "Bar" type
export * from './v2.ts'

// We could do explicit exports from "latest". Same as type alias approach, but with a fewer lines of code.

export { Foo, GetFooHTTP } from './v2.ts'
export { Bar } from './v1.ts'

```

WDYT @pgayvallet and @Rudolf ? I was thinking we should write down a "How to version TS interfaces guide" for teams to reference - perhaps as a follow up task to merging this PR?

---

> So overall I'm not sure yet the upsides outweights the downsides [of using `extends` on TS interfaces]?

Yeah, good point. I had not considered the generated/extracted docs. Two things I like about the `extends` pattern:

1. It preserves history perfectly: ancestors are all linked so no fields will go missing. Perhaps the plus-side of the negative UX you identified.
2. You only document fields once

I hear what you're saying about indirection, it is useful to see all fields represented in an interface. To get (2) without `extends` we can leverage `@interhitdoc`. I'm not sure how we'd avoid the (small) risk in (1) of human error. Perhaps it is negligible. Happy to capture this in a guide and iterate on it further there.

--- 

@Rudolf 

> But this pattern can work even with explicitly mentioning all the types, it's just mork work to keep in sync.

Not 100% following. Are you referring to the the `import { Foo as FooV1 } './v1';` pattern? Or are you talking about the explicit imports I mentioned above?

Reply:
> I assume that interfaces won't necessarily all have new versions at the same rate

Yeah, that may be a fair point. 

From where I take my examples (k8s), this isn't an issue, given version types are autogenerated, therefor all the 'types' are present in all the version, even if no changes are present. Without codegen, it will probably not be the case, and we may not want to manually 'copy-paste' a type.

Still, I think that ideally, we should be able to import all the types for a given version from a single given package / folder, even if the types didn't change.

So imagine that `Bar` didn't evolve between v1 and v2, we could import it from `api/v1` and `api/v2`.

Now, it's all about defining what the correct granularity for 'versions' is, I suppose... Do we want a single version number of those versioned types for a given API facade, or is the version specific to each individual API, or each individual type... I honestly have no idea...